airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStreamCopier.java

```diff
@@ -27,24 +40,90 @@ public RedshiftStreamCopier(String stagingFolder,
+    possibleManifest.stream()
+        .map(this::putManifest)
+        .forEach(this::executeCopy);
+    LOGGER.info("Copy to tmp table {} in destination for stream {} complete.", tmpTableName, streamName);
   }
 
   @Override
-  public void copyS3CsvFileIntoTable(JdbcDatabase database, String s3FileLocation, String schema, String tableName, S3Config s3Config)
-      throws SQLException {
+  public void copyS3CsvFileIntoTable(JdbcDatabase database, String s3FileLocation, String schema, String tableName, S3Config s3Config) {
+    throw new RuntimeException("Redshift Stream Copier should not copy individual files without use of a manifest");
+  }
+
+  @Override
+  public void removeFileAndDropTmpTable() throws Exception {
+    super.removeFileAndDropTmpTable();
+    if (manifestFilePath != null) {
+      LOGGER.info("Begin cleaning s3 manifest file {}.", manifestFilePath);
+      if (s3Client.doesObjectExist(s3Config.getBucketName(), manifestFilePath)) {
+        s3Client.deleteObject(s3Config.getBucketName(), manifestFilePath);
+      }
+      LOGGER.info("S3 manifest file {} cleaned.", manifestFilePath);
+    }
+  }
+
+  /**
+   * Creates the contents of a manifest file given the `s3StagingFiles`. There must be at least one
+   * entry in a manifest file otherwise it is not considered valid for the COPY command.
+   *
+   * @return null if no stagingFiles exist otherwise the manifest body String
+   */
+  private String createManifest() {
+    if (s3StagingFiles.isEmpty()) {
+      return null;
+    }
+
+    final var s3FileEntries = s3StagingFiles.stream()
+        .map(filePath -> new Entry(getFullS3Path(s3Config.getBucketName(), filePath)))
+        .collect(Collectors.toList());
+    final var manifest = new Manifest(s3FileEntries);
+
+    return Exceptions.toRuntime(() -> objectMapper.writeValueAsString(manifest));
+  }
+
+  /**
+   * Upload the supplied manifest file to S3
+   *
+   * @param manifestContents the manifest contents, never null
+   * @return the path where the manifest file was placed in S3
+   */
+  private String putManifest(String manifestContents) {
+    manifestFilePath =
+        String.join("/", stagingFolder, schemaName, String.format("%s.manifest", UUID.randomUUID()));
+
+    s3Client.putObject(s3Config.getBucketName(), manifestFilePath, manifestContents);
+
+    return manifestFilePath;
+  }
+
+  /**
+   * Run Redshift COPY command with the given manifest file
+   *
+   * @param manifestPath the path in S3 to the manifest file
+   */
+  private void executeCopy(String manifestPath) {
     final var copyQuery = String.format(
         "COPY %s.%s FROM '%s'\n"
             + "CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'\n"
             + "CSV REGION '%s' TIMEFORMAT 'auto'\n"
-            + "STATUPDATE OFF;\n",
-        schema,
-        tableName,
-        s3FileLocation,
+            + "STATUPDATE OFF\n"
+            + "MANIFEST;",
```

Comment:
Hi, @archaean. I only have one question. I deliberately split one temporary file on S3 or GCS for every stream into several smaller files to avoid a timeout exception in a copy query from a file to a temp table. This happened when the work was done on a large amount of data. Are you sure using the manifest won't get us back to this issue?

Reply:
@andriikorotkov , I guess there are edge cases where it could, but the data would likely need to be an order of magnitude larger than whatever caused the timeout in your case given the parallelism.

I don't know exactly what timeout exception you are referring to, but if it from the Redshift connection, it is likely `ERROR:  Query (150) canceled on user's request` which is generally caused when a query reaches the timeout limit in redshift.  I assume the COPY command is also affected by this query limitation, but this timeout is [configurable](https://docs.aws.amazon.com/redshift/latest/dg/r_statement_timeout.html).

[Other timeouts](https://docs.aws.amazon.com/redshift/latest/dg/queries-troubleshooting.html#queries-troubleshooting-query-hangs) are possible for long running COPY commands due to firewall configurations.

One possible implementation would be to further batch the files to arbitrary number (hopefully close to the number of shards on the redshift cluster) and make a manifest file for each batch and do serial copy commands that way. 

So to answer your question, it won't be the same issue because the copy command should process an order of magnitude faster than serial copy commands depending on the configuration of the Redshift cluster.  *I think* the query timeout limit can be configured to avoid this issue.   And while batching the files into multiple manifest and copy commands is an option, at this point it seems like a premature optimization.